### PR TITLE
Homework6

### PR DESCRIPTION
**Pull request by AndreReen**
HW6 ready for review

### CODE REVIEW

File: src/test/java/ru/training/at/hw4/potest/TestExerciseSteps2.java [hunk 1]
```diff
@@ -5,8 +5,7 @@
 import org.testng.annotations.Parameters;
 import org.testng.annotations.Test;
 import ru.training.at.hw4.steps.DifferentElementsSteps;
-import ru.training.at.hw4.steps.MainPageSteps;
-import ru.training.at.hw4.testdata.TestData;
+
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw4/potest/TestExerciseSteps2.java:9]: Files from previous HW should not present in this PR

File: src/test/java/ru/training/at/hw6/Actions/Actions.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import ru.training.at.hw6.entities.User;
+import static ru.training.at.hw6.PageObjects.JdiSite.*;
+
+public class Actions {
+
+    @Step
+    public static void login() {
+        userIcon.click();
+        loginForm.submit(new User(), "enter");
+    }
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw6/Actions/Actions.java:14]: the method should accept some Entity as an argument

File: src/test/java/ru/training/at/hw6/Actions/CreateDataProvider.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
+package ru.training.at.hw6.Actions;
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw6/Actions/CreateDataProvider.java:1]: The class should be in some other package (actions is a wrong place)

\+ package name should not contain capital cased letters

File: src/test/java/ru/training/at/hw6/PageObjects/JdiSite.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+import org.openqa.selenium.WebElement;
+import ru.training.at.hw6.forms.LoginForm;
+import ru.training.at.hw6.forms.MetalsAndColorsForm;
+
+@JSite("https://jdi-testing.github.io/jdi-light/")
+public class JdiSite {
+    public static HomePage homePage;
+    public static MetalsAndColorsPage macPage;
+
+    @UI(".form") public static MetalsAndColorsForm macForm;
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw6/PageObjects/JdiSite.java:18]: not obvious name

File: src/test/java/ru/training/at/hw6/PageObjects/MetalsAndColorsPage.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    @UI(".summ-res")
+    public static Text summRes;
+    @UI(".elem-res")
+    public static Text elemRes;
+    @UI(".col-res")
+    public static Text colorRes;
+    @UI(".met-res")
+    public static Text metalRes;
+    @UI(".sal-res")
+    public static Text vegRes;
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw6/PageObjects/MetalsAndColorsPage.java:21]: This approach can be simplified by using List

File: src/test/java/ru/training/at/hw6/entities/TestValues.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package ru.training.at.hw6.entities;
+
+import lombok.Data;
+
+@Data
+public class TestValues {
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw6/entities/TestValues.java:6]: The class name should reflect its purpose

File: src/test/java/ru/training/at/hw6/forms/MetalsAndColorsForm.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+import com.epam.jdi.light.elements.complex.Checklist;
+import com.epam.jdi.light.elements.complex.dropdown.Dropdown;
+import com.epam.jdi.light.elements.composite.Form;
+import com.epam.jdi.light.elements.pageobjects.annotations.locators.JDropdown;
+import com.epam.jdi.light.elements.pageobjects.annotations.locators.UI;
+import com.epam.jdi.light.ui.html.elements.common.Button;
+import com.epam.jdi.light.ui.html.elements.complex.RadioButtons;
+import ru.training.at.hw6.entities.TestValues;
+
+public class MetalsAndColorsForm extends Form<MetalsAndColorsForm> {
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw6/forms/MetalsAndColorsForm.java:13]: This class should extend Form with DataClass
For the example check `LoginForm`

File: src/test/java/ru/training/at/hw6/forms/MetalsAndColorsForm.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+            list = "li",
+            expand = ".caret")
+    public Dropdown vegetables;
+
+    @UI("#submit-button")
+    public Button submitButton;
+
+    public void submin() {
+        submitButton.click();
+    }
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw6/forms/MetalsAndColorsForm.java:47]: This method can be overridden with DataClass as argument

File: src/test/java/ru/training/at/hw6/forms/MetalsAndColorsForm.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    public void fillForm(TestValues data) {
+        summaryOdd.select(data.getSummary()[0]);
+        summaryEven.select(data.getSummary()[1]);
+        elements.select(data.getElements());
+        colors.select(data.getColor());
+        metals.select(data.getMetals());
+        vegetables.expand();
+        vegetables.select(3);
+        vegetables.select(data.getVegetables());
+    }
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw6/forms/MetalsAndColorsForm.java:58]: Get rid of this method in favor of `submit` override

File: src/test/java/ru/training/at/hw6/tests/TheTest.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+import org.testng.Assert;
+import org.testng.annotations.Test;
+import ru.training.at.hw6.Actions.Actions;
+import ru.training.at.hw6.Actions.CreateDataProvider;
+import ru.training.at.hw6.entities.TestValues;
+import java.util.Arrays;
+import static ru.training.at.hw6.PageObjects.JdiSite.*;
+import static ru.training.at.hw6.PageObjects.MetalsAndColorsPage.*;
+
+public class TheTest implements TestsInit {
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw6/tests/TheTest.java:12]: Test class name should reflect its purpose

File: src/test/java/ru/training/at/hw6/tests/TheTest.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+    @Test(dataProviderClass = CreateDataProvider.class, dataProvider = "dataFromFile")
+    public void MetalsColorsFormFillTest(TestValues data) {
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw6/tests/TheTest.java:15]: 1. Code convention breaker
1.  What do you mean by `data` argument name?

File: src/test/java/ru/training/at/hw6/Actions/CreateDataProvider.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+import com.google.gson.stream.JsonReader;
+import org.testng.annotations.DataProvider;
+import ru.training.at.hw6.entities.TestValues;
+
+import java.io.FileReader;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.*;
+
+public class CreateDataProvider {
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw6/Actions/CreateDataProvider.java:14]: Name should have some meaning

File: src/test/java/ru/training/at/hw6/tests/TheTest.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+        homePage.open();
+        Actions.login();
+        headerMenu.select("METALS & COLORS");
+        macForm.fillForm(data);
+        macForm.submin();
+        verifyResults(data);
+        Actions.logout();
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw6/tests/TheTest.java:22]: Multiple styles (actions, page object actions straight calls and some method inside test class `verifyResults`)

File: src/test/java/ru/training/at/hw6/tests/TheTest.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+import static ru.training.at.hw6.PageObjects.JdiSite.*;
+import static ru.training.at.hw6.PageObjects.MetalsAndColorsPage.*;
+
+public class TheTest implements TestsInit {
+
+    @Test(dataProviderClass = CreateDataProvider.class, dataProvider = "dataFromFile")
+    public void MetalsColorsFormFillTest(TestValues data) {
+        homePage.open();
+        Actions.login();
+        headerMenu.select("METALS & COLORS");
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw6/tests/TheTest.java:18]: Pure string is not a good choice, consider use of constants of Enum

File: src/test/java/ru/training/at/hw6/tests/TheTest.java [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+        String colorExp = "Color: " + data.getColor();
+        String metalExp = "Metal: " + data.getMetals();
+        String vegBuild = Arrays.toString(data.getVegetables());
+        String vegExp = "Vegetables: " + vegBuild.substring(1, vegBuild.length() - 1);
+
+        Assert.assertEquals(summRes.getText(), sumExp);
+        Assert.assertEquals(elemRes.getText(), elemExp);
+        Assert.assertEquals(colorRes.getText(), colorExp);
+        Assert.assertEquals(metalRes.getText(), metalExp);
+        Assert.assertEquals(vegRes.getText(), vegExp);
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw6/tests/TheTest.java:38]: This approach can be simplified, compare two collections
`List<> expected, List<> actualValuesFromElement`

File: src/test/java/ru/training/at/hw6/tests/TheTest.java [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    public void verifyResults(TestValues data) {
+        String sumExp = "Summary: " + (Integer.parseInt(data.getSummary()[0]) + Integer.parseInt(data.getSummary()[1]));
+        String elemBuild = Arrays.toString(data.getElements());
+        String elemExp = "Elements: " + elemBuild.substring(1, elemBuild.length() - 1);
+        String colorExp = "Color: " + data.getColor();
+        String metalExp = "Metal: " + data.getMetals();
+        String vegBuild = Arrays.toString(data.getVegetables());
+        String vegExp = "Vegetables: " + vegBuild.substring(1, vegBuild.length() - 1);
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw6/tests/TheTest.java:32]: This method should not be in this class, it breaks one of the design principles

**Review by shoggy — CHANGES_REQUESTED** [review]
putting annotation and variable definition in one line makes code not readable

**Comment by shoggy** [comment]
Please modify checkstyle settings, copy-paste it from [example](https://github.com/DmitryKhodakovsky/epam-test-auto-spring-2021/blob/master/pom.xml) (plugin definition and versions)

File: src/test/java/ru/training/at/hw6/Actions/CreateDataProvider.java (hunk 3 above)
Comment (AndreReen) [reply src/test/java/ru/training/at/hw6/Actions/CreateDataProvider.java:1]: Moved class to data_providers package.
Updated package names to meet naming conventions.

File: src/test/java/ru/training/at/hw6/Actions/Actions.java (hunk 2 above)
Comment (AndreReen) [reply src/test/java/ru/training/at/hw6/Actions/Actions.java:14]: Updated

File: src/test/java/ru/training/at/hw6/PageObjects/JdiSite.java (hunk 4 above)
Comment (AndreReen) [reply src/test/java/ru/training/at/hw6/PageObjects/JdiSite.java:18]: Updated name to be more clear =  metalAndColorForm

File: src/test/java/ru/training/at/hw6/PageObjects/MetalsAndColorsPage.java (hunk 5 above)
Comment (AndreReen) [reply src/test/java/ru/training/at/hw6/PageObjects/MetalsAndColorsPage.java:21]: Implemented list of Webelements

File: src/test/java/ru/training/at/hw4/potest/TestExerciseSteps2.java (hunk 1 above)
Comment (AndreReen) [reply src/test/java/ru/training/at/hw4/potest/TestExerciseSteps2.java:9]: Crossover from refactoring previous HW

File: src/test/java/ru/training/at/hw6/entities/TestValues.java (hunk 6 above)
Comment (AndreReen) [reply src/test/java/ru/training/at/hw6/entities/TestValues.java:6]: Updated

File: src/test/java/ru/training/at/hw6/forms/MetalsAndColorsForm.java (hunk 7 above)
Comment (AndreReen) [reply src/test/java/ru/training/at/hw6/forms/MetalsAndColorsForm.java:13]: Fixed

File: src/test/java/ru/training/at/hw6/forms/MetalsAndColorsForm.java (hunk 9 above)
Comment (AndreReen) [reply src/test/java/ru/training/at/hw6/forms/MetalsAndColorsForm.java:58]: Done

File: src/test/java/ru/training/at/hw6/tests/TheTest.java (hunk 10 above)
Comment (AndreReen) [reply src/test/java/ru/training/at/hw6/tests/TheTest.java:12]: Updated

File: src/test/java/ru/training/at/hw6/Actions/CreateDataProvider.java (hunk 12 above)
Comment (AndreReen) [reply src/test/java/ru/training/at/hw6/Actions/CreateDataProvider.java:14]: Updated

File: src/test/java/ru/training/at/hw6/tests/TheTest.java (hunk 14 above)
Comment (AndreReen) [reply src/test/java/ru/training/at/hw6/tests/TheTest.java:18]: Created enum class with header menu items

File: src/test/java/ru/training/at/hw6/tests/TheTest.java (hunk 16 above)
Comment (AndreReen) [reply src/test/java/ru/training/at/hw6/tests/TheTest.java:32]: Made this method a part of MetalsAndColors Entity

File: src/test/java/ru/training/at/hw6/tests/TheTest.java (hunk 15 above)
Comment (AndreReen) [reply src/test/java/ru/training/at/hw6/tests/TheTest.java:38]: Implemented

File: src/test/java/ru/training/at/hw6/tests/TheTest.java (hunk 14 above)
Comment (AndreReen) [reply src/test/java/ru/training/at/hw6/tests/TheTest.java:18]: Created enum for  header menu items.

File: src/test/java/ru/training/at/hw6/Actions/CreateDataProvider.java (hunk 12 above)
Comment (AndreReen) [reply src/test/java/ru/training/at/hw6/Actions/CreateDataProvider.java:14]: Updated to DataProviderForMetalAndColorForm.java

File: src/test/java/ru/training/at/hw6/PageObjects/MetalsAndColorsPage.java (hunk 5 above)
Comment (AndreReen) [reply src/test/java/ru/training/at/hw6/PageObjects/MetalsAndColorsPage.java:21]: Implemented

File: src/test/java/ru/training/at/hw6/PageObjects/JdiSite.java (hunk 4 above)
Comment (AndreReen) [reply src/test/java/ru/training/at/hw6/PageObjects/JdiSite.java:18]: Updated to metalAndColorForm

File: src/test/java/ru/training/at/hw6/tests/TheTest.java (hunk 13 above)
Comment (AndreReen) [reply src/test/java/ru/training/at/hw6/tests/TheTest.java]: Refactored to Actions style

**Review by shoggy — APPROVED** [review]

**Pull request merged**